File: csharp-project/RekindleBookStoreAqa/RekindleTaf/Tests/Api/CustomerServiceTest.cs
using System;
using System.Collections.Generic;
using System.Net;
using FluentAssertions;
using NUnit.Extension.DependencyInjection;
using NUnit.Framework;
using NUnit.Framework.Internal;
using RekindleTaf.Configuration;
using RekindleTaf.Configuration.Rest;
using RekindleTaf.Controllers;
using RekindleTaf.Exceptions;
using RekindleTaf.Model;
using RekindleTaf.Model.Factory;
using RekindleTaf.Utils;
using RestSharp;

namespace RekindleTaf.Tests.Api;

[DependencyInjectingTestFixture]
public class CustomerServiceTest(IFactoryKit factoryKit, ICustomerController customerController) : Base
{
    private readonly List<Customer> Customers = new();

    [OneTimeTearDown]
    public void Cleanup()
    {
        if (Customers.Count == 0) return;
        Customers.ForEach(customerController.DeleteSuccessfully);
    }
    
    [Test]
    public void Create_NewUser_Should_Return_OK()
    {
        var customer = factoryKit.Factory<Customer>(FactoryKey.Customer).Create(CustomerKey.StdUser);
        customer = customerController.PostSuccessfully(customer);
        Customers.Add(customer);
        customer.Id.Should().NotBeEmpty();
        customer.FirstName.Should().NotBeEmpty();
        customer.LastName.Should().NotBeEmpty();
        customer.Username.Should().NotBeEmpty();
    }
    
    [Test]
    public void Get_NewUser_Should_Return_OK()
    {
        var customer = factoryKit.Factory<Customer>(FactoryKey.Customer).Create(CustomerKey.StdUser);
        customer = customerController.PostSuccessfully(customer);
        Customers.Add(customer);
        var fetchedCustomer = customerController.GetSuccessfully(customer.Id!);
        fetchedCustomer.Should().BeEquivalentTo(customer);
    }
    
    [Test]
    public void Get_AllCustomers_Should_Return_OK()
    {
        var fetchedCustomers = customerController.GetAllSuccessfully();
        fetchedCustomers.Should().NotBeEmpty();
    }
    
    [Test]
    public void Update_NewUser_Should_Return_NoContent()
    {
        var customer = factoryKit.Factory<Customer>(FactoryKey.Customer).Create(CustomerKey.StdUser);
        customer = customerController.PostSuccessfully(customer);
        Customers.Add(customer);
        var updatedUsername = "Updated Username";
        customer.Username = updatedUsername;
        customerController.PutSuccessfully(customer);
        var fetchedCustomer = customerController.GetSuccessfully(customer.Id!);
        fetchedCustomer.Username.Should().Be(updatedUsername);
    }
    
    [Test]
    public void Delete_NewUser_Should_Return_OK()
    {
        var customer = factoryKit.Factory<Customer>(FactoryKey.Customer).Create(CustomerKey.StdUser);
        customer = customerController.PostSuccessfully(customer);
        customerController.DeleteSuccessfully(customer);
        var response = customerController.Get(customer.Id!).Result;
        response.StatusCode.Should().Be(HttpStatusCode.BadRequest);
        response.GetNode("errorMessage").Should().Be(ErrorMessage.CustomerNotFound);
    }
}

File: typescript-project/rekindle-book-store-aqa/tests/customer-service.test.ts
import {afterAll, beforeAll, describe, expect, test} from "@jest/globals";
import {Customer} from "@/model/domain-model";
import {Factory, FactoryKey, FactoryKit} from "@/model/factory";
import {CustomerController} from "@/rest/controllers/customer-controller";
import {CustomerKey} from "@/model/customer-factory";
import {HttpStatusCode} from "axios";
import {ErrorMessage} from "../src/utils/utils";
import {container} from "tsyringe";
import {RestClient} from "@/rest/rekindle-rest-client";

describe('Customer Service Test', () => {
    let restClient : RestClient = container.resolve('RestClient');
    let factoryKit: FactoryKit = container.resolve('FactoryKit');
    let customerController: CustomerController = container.resolve('CustomerCrudController');
    let customerFactory : Factory<Customer> = factoryKit.factory<Customer>(FactoryKey.customer);
    let customers: Customer[] = [];

    beforeAll(async () => {
        await restClient.setupToken();
    });

    afterAll(async () => {
        if (customers.length === 0) return;
        for (const customer of customers) {
            await customerController.deleteSuccessfully(customer);
        }
    });

    test('should create a new user and return OK', async () => {
        let customer = customerFactory.create(CustomerKey.stdCustomerActive);
        customer = await customerController.postSuccessfully(customer);
        customers.push(customer);

        expect(customer.id).toBeDefined();
        expect(customer.firstName).toBeTruthy();
        expect(customer.lastName).toBeTruthy();
        expect(customer.username).toBeTruthy();
    });

    test('should get a new user and return OK', async () => {
        let customer = customerFactory.create(CustomerKey.stdCustomerActive);
        customer = await customerController.postSuccessfully(customer);
        customers.push(customer);
        const fetchedCustomer = await customerController.getSuccessfully(customer.id);
        expect(fetchedCustomer).toEqual(customer);
    });

    test('should get all customers and return OK', async () => {
        const fetchedCustomers = await customerController.getAllSuccessfully();
        expect(fetchedCustomers).not.toHaveLength(0);
    });

    test('should update a new user and return no content', async () => {
        let customer = customerFactory.create(CustomerKey.stdCustomerActive);
        customer = await customerController.postSuccessfully(customer);
        customers.push(customer);

        const updatedUsername = "Updated Username";
        customer.username = updatedUsername;
        await customerController.putSuccessfully(customer);

        const fetchedCustomer = await customerController.getSuccessfully(customer.id!);
        expect(fetchedCustomer.username).toBe(updatedUsername);
    });

    test('should delete a new user and return OK', async () => {
        let customer = customerFactory.create(CustomerKey.stdCustomerActive);
        customer = await customerController.postSuccessfully(customer);
        await customerController.deleteSuccessfully(customer);

        await customerController.get(customer.id!)
            .catch((error) => {
                expect(error.response.status).toBe(HttpStatusCode.BadRequest);
                expect(error.response.data.errorMessage).toBe(ErrorMessage.customerNotFound);
            })
    });
});